helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -66,29 +66,26 @@
   @Override
   public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebalanceException {
     OptimalAssignment optimalAssignment = new OptimalAssignment();
-    Map<String, Set<AssignableReplica>> replicasByResource = clusterModel.getAssignableReplicaMap();
     List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
-
-    // TODO: different orders of resource/replica could lead to different greedy assignments, will
-    // revisit and improve the performance
-    for (String resource : replicasByResource.keySet()) {
-      for (AssignableReplica replica : replicasByResource.get(resource)) {
-        Optional<AssignableNode> maybeBestNode =
-            getNodeWithHighestPoints(replica, nodes, clusterModel.getContext(), optimalAssignment);
-        // stop immediately if any replica cannot find best assignable node
-        if (optimalAssignment.hasAnyFailure()) {
-          String errorMessage = String.format(
-              "Unable to find any available candidate node for partition %s; Fail reasons: %s",
-              replica.getPartitionName(), optimalAssignment.getFailures());
-          throw new HelixRebalanceException(errorMessage,
-              HelixRebalanceException.Type.FAILED_TO_CALCULATE);
-        }
-        maybeBestNode.ifPresent(node -> clusterModel.assign(replica.getResourceName(),
-            replica.getPartitionName(), replica.getReplicaState(), node.getInstanceName()));
+    // Sort the replicas so the input is stable for the greedy algorithm.
+    // For the other algorithm implementation, this sorting could be unnecessary.
+    for (AssignableReplica replica : getOrderedAssignableReplica(clusterModel)) {
```

Comment:
Should we add some pre-processing step that does whatever is necessary for the input? 
For example, we may require different pre-processing. In this case, for the greedy algo, we need to sort the replicas. For other algos, we may need different types of pre-processing (such as sort by fault zones, etc.).

Reply:
This belongs to the algorithm implementation detail. The rebalaner ensures the input contains enough data for the algorithm to process.

Reply:
This doesn't answer my question...

Reply:
The answer is no. Even we have the "pre-processing" step, the rebalancer shall not be aware of that. On the other hand, for the constraint-based algorithm class, it is not necessary to split the ordering logic as a different step for now.
However, with @i3wangyi 's tuning work, we might need to refine it further.